README.md

```diff
@@ -1,3 +1,317 @@
+```python
+responses.PaginatedResponse(some_queryset, SomeDRFSerializer, request)
+```
+will generate:
+```
+{
+    "objects": [{id: 1}, {id: 2}, {id: 3}, ... ],
+    "pagination": {
+        "page": 1,
+        "page_size": 20,
+        "num_pages": 5,
+        "total_objects": 89,
+        "next_page": "/api/v1/orders?page=2&page_size=20",
+        "previous_page": null
+    }
+}
+```
+Alternatively, instead of a DRF Serializer class, you can pass in a function that takes in a list of objects and returns a list of serialized objects,
+```python
+responses.PaginatedResponse(
+    some_queryset,
+    lambda objs: [obj.to_dict() for obj in objs],
+    request,
+)
+```
+See the `Pagination` section for more details.
+
+### Error Responses
+
+All error responses (4xx) will have the following shape:
+```json
+{"error_code": "some_unique_error_string", "detail": "More details about the error..."}
+```
+Some allow the `error_code` and `detail` to be specified, while others have them hard-coded.
+
+#### `UnauthorizedResponse`
+* 403
+* `error_code`
+	* `"unauthorized"`
+* `detail`
+	* `"Unauthorized to access this resource"`
+
+#### `NotFoundResponse`
+* 404
+* `error_code`
+	* `"not_found"`
+* `detail`
+	* `"Resource not found"`
+
+#### `ClientErrorResponse`
+* 400
+* To be used when the client made an error it could have avoided.
+* `error_code` and `detail` must be specified.
+```python
+responses.ClientErrorResponse("quote_state_error", "Cannot approved an already approved quote")
+```
+
+#### `ServerValidationErrorResponse`
+* 422
+* To be used when a validation error occurs that could only be detected by the server.
+* `error_code` and `detail` must be specified.
+ ```python
+responses.ClientErrorResponse("email_already_taken", "The email you have provided is already in use")
+```
+
+## Exceptions
+
+### ValidationError
+Use sparingly. If raised, the framework will catch it and return generate a `ClientErrorResponse`, with `"validation_error"` as the `error_code` and the error message as `detail`. However, in most cases you should just catch your own errors and return `ClientErrorResponse`.
+
+### ImmediateResponseError
+This exception takes in a response instance and, when raised, the framework will catch it and return the response it was instantiated with. This is useful when you have some shared function between two view functions and want to quickly bail in the shared function.
+```python
+class SomeObjView(View):
+    def _get_some_obj(self, id):
+        try:
+            return SomeObj.objects.get(id=id)
+        except:
+            raise exceptions.ImmediateResponseError(responses.NotFoundResponse())
+
+    @view.method("GET")
+    def get_some_obj(request, id=None):
+        obj = self._get_some_obj(id)
+        ...
+
+    @view.method("PATCH")
+    def patch_some_obj(request, id=None):
+        obj = self._get_some_obj(id)
+        ...
+```
+## Schemas & Fields
+All schema fields support the following arguments:
```

Comment:
Would it be beneficial to mention that the schemas and fields are powered by `marshmallow`? In case anyone wants to add any field that we may have missed?

Reply:
Ohhh yes